tests/chainer_tests/functions_tests/math_tests/test_matmul.py

```diff
@@ -46,230 +53,99 @@ def test_matmul_backward_gpu(self):
-            numpy.dot(self.x1[i], self.x2)
-            for i in six.moves.range(batch_size)])
+@testing.parameterize(*testing.product_dict(
+    [
+        # matmul
+        {'x1_shape': (2, 5), 'x2_shape': (5, 10), 'gy_shape': (2, 10),
+         'transa': False, 'transb': False},
+        {'x1_shape': (5, 2), 'x2_shape': (5, 10), 'gy_shape': (2, 10),
+         'transa': True, 'transb': False},
+        {'x1_shape': (2, 5), 'x2_shape': (10, 5), 'gy_shape': (2, 10),
+         'transa': False, 'transb': True},
+        {'x1_shape': (5, 2), 'x2_shape': (10, 5), 'gy_shape': (2, 10),
+         'transa': True, 'transb': True},
+
+        # vector
+        {'x1_shape': (5,), 'x2_shape': (5,), 'gy_shape': (),
+         'transa': True, 'transb': False},
+        {'x1_shape': (5,), 'x2_shape': (5,), 'gy_shape': (),
+         'transa': False, 'transb': True},
+
+        # batched matmul
+        {'x1_shape': (6, 2, 5), 'x2_shape': (6, 5, 10), 'gy_shape': (6, 2, 10),
+         'transa': False, 'transb': False},
+        {'x1_shape': (6, 5, 2), 'x2_shape': (6, 5, 10), 'gy_shape': (6, 2, 10),
+         'transa': True, 'transb': False},
+        {'x1_shape': (6, 2, 5), 'x2_shape': (6, 10, 5), 'gy_shape': (6, 2, 10),
+         'transa': False, 'transb': True},
+        {'x1_shape': (6, 5, 2), 'x2_shape': (6, 10, 5), 'gy_shape': (6, 2, 10),
+         'transa': True, 'transb': True},
+
+        # batchsize = 1
+        {'x1_shape': (1, 2, 5), 'x2_shape': (1, 5, 10), 'gy_shape': (1, 2, 10),
+         'transa': False, 'transb': False},
+
+        # 4dim batched matmul
+        {'x1_shape': (2, 3, 4, 5), 'x2_shape': (2, 3, 5, 6),
+         'gy_shape': (2, 3, 4, 6), 'transa': False, 'transb': False},
+    ],
+    [
+        {'x1_dtype': numpy.float16},
+        {'x1_dtype': numpy.float32},
+        {'x1_dtype': numpy.float64},
+    ],
+    [
+        {'x2_dtype': numpy.float16},
+        {'x2_dtype': numpy.float32},
+        {'x2_dtype': numpy.float64},
+    ]
+))
+class TestMatMul(_TestMatMul):
+
+    def setUp(self):
+        self.x1 = numpy.random.uniform(.5, 1, self.x1_shape)
+        self.x1 = self.x1.astype(self.x1_dtype)
+        self.x2 = numpy.random.uniform(.5, 1, self.x2_shape)
+        self.x2 = self.x2.astype(self.x2_dtype)
+        ret_dtype = numpy.result_type(self.x1_dtype, self.x2_dtype)
+        self.gy = numpy.random.uniform(-1, 1, self.gy_shape).astype(ret_dtype)
+
+        self.op = lambda x, y: F.matmul(x, y, transa=self.transa,
+                                        transb=self.transb)
+        if self.x1.ndim == 1:
+            x1 = self.x1
+        else:
+            x1 = self.x1.swapaxes(-1, -2) if self.transa else self.x1
+
+        if self.x2.ndim == 1:
+            x2 = self.x2
+        else:
+            x2 = self.x2.swapaxes(-1, -2) if self.transb else self.x2
+        if x1.ndim <= 2:
+            self.forward_answer = numpy.dot(x1, x2)
+        else:
+            self.forward_answer = numpy.einsum('...ij,...jk->...ik', x1, x2)
```

Comment:
Please factor out from line 115 (`if self.x1.ndim == 1:`) to line 127 (`self.forward_answer = numpy.einsum('...ij,...jk->...ik', x1, x2)`) as a method, something like:
```
def _get_forward_answer(self, x1, x2, transa, transb):
    ...
```